src/AppInstallerCLICore/Workflows/UninstallFlow.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "UninstallFlow.h"
+#include "WorkflowBase.h"
+#include "ShellExecuteInstallerHandler.h"
+#include "AppInstallerMsixInfo.h"
+
+using namespace AppInstaller::Manifest;
+using namespace AppInstaller::Msix;
+using namespace AppInstaller::Repository;
+
+namespace AppInstaller::CLI::Workflow
+{
+    void GetUninstallInfo(Execution::Context& context)
+    {
+        auto installedPackageVersion = context.Get<Execution::Data::InstalledPackageVersion>();
+        const std::string installedTypeString = installedPackageVersion->GetMetadata()[PackageVersionMetadata::InstalledType];
+        switch (ManifestInstaller::ConvertToInstallerTypeEnum(installedTypeString))
+        {
+        case ManifestInstaller::InstallerTypeEnum::Exe:
+        case ManifestInstaller::InstallerTypeEnum::Burn:
+        case ManifestInstaller::InstallerTypeEnum::Inno:
+        case ManifestInstaller::InstallerTypeEnum::Nullsoft:
+        {
+            IPackageVersion::Metadata packageMetadata = installedPackageVersion->GetMetadata();
+
+            // Default to silent unless it is not present or interactivity is requested
+            auto uninstallCommandItr = packageMetadata.find(PackageVersionMetadata::SilentUninstallCommand);
+            if (uninstallCommandItr == packageMetadata.end() || context.Args.Contains(Execution::Args::Type::Interactive))
+            {
+                auto interactiveItr = packageMetadata.find(PackageVersionMetadata::StandardUninstallCommand);
+                if (interactiveItr != packageMetadata.end())
+                {
+                    uninstallCommandItr = interactiveItr;
+                }
+            }
+
+            if (uninstallCommandItr == packageMetadata.end())
+            {
+                context.Reporter.Error() << Resource::String::NoUninstallInfoFound << std::endl;
+                AICLI_TERMINATE_CONTEXT(APPINSTALLER_CLI_ERROR_NO_UNINSTALL_INFO_FOUND);
+            }
+
+            context.Add<Execution::Data::UninstallString>(uninstallCommandItr->second);
+            break;
+        }
+        case ManifestInstaller::InstallerTypeEnum::Msi:
+        case ManifestInstaller::InstallerTypeEnum::Wix:
+        {
+            // Uninstall strings for MSI don't include UI level (/q) needed to avoid interactivity,
+            // so we handle them differently.
+            auto productCodes = installedPackageVersion->GetMultiProperty(PackageVersionMultiProperty::ProductCode);
+            if (productCodes.empty())
```

Comment:
>if (productCodes.empty()) [](start = 12, length = 25)

I believe in current state we'll have to fall back to uninstallstring here.
The installedPackageVersion->GetMultiProperty() is a method for querying index. This will only work iff the installed package is correlated back to an entry in index and the manifest in the index contains productcode declarations.

Or if the only reason we do this is /q not in the uninstall string, can we add /q by default unless it's /interactive mode?

Reply:
Also do you know if it's common that ISV sets custom args? if we construct own own from product code, I'm afraid custom args will be lost, if there exists one

---
In reply to: [543052441](https://github.com/microsoft/winget-cli/pull/659#discussion_r543052441) [](ancestors = 543052441)

Reply:
> The installedPackageVersion->GetMultiProperty() is a method for querying index. This will only work iff the installed package is correlated back to an entry in index

I believe this should always come from the Installed source which always sets the product code on `ARPHelper::PopulateIndexFromKey`. If we have the uninstall string then we necessarily also have the product code from the ARP key name.

> Or if the only reason we do this is /q not in the uninstall string, can we add /q by default unless it's /interactive mode?

I considered it, but to add /q we need to either be completely sure it is not possible to already have a /q flag here, be sure that the flag we add at the end will always override any flag that may already be there, or parse the string to see if it has one to remove it. That all seems too complicated and I'm pretty sure that constructing the command ourselves works fine because that's what the Settings app, and Programs and Features do for uninstalling.

> Also do you know if it's common that ISV sets custom args? if we construct own own from product code, I'm afraid custom args will be lost, if there exists one

Yes, it would be lost but I couldn't find anything about how to set custom args in an MSI, so I'm not even sure if it is possible without going around what the MSI provides. I know it's not proof, but anecdotically on my machines no entry for MSI had any custom args.

Reply:
Ah, I overlooked the index.addmanifest line in that method. This makes sense.

---
In reply to: [543567716](https://github.com/microsoft/winget-cli/pull/659#discussion_r543567716) [](ancestors = 543567716)